core/state/statedb.go

```diff
@@ -973,24 +982,70 @@ func (db *DB) AddReward(snapshot *stk.ValidatorWrapper, reward *big.Int, shareLo
 		rewardPool.Sub(rewardPool, commissionInt)
 	}
 
+	// no sanity check for length of delegations between curValidator and snapshot
+	// if a delegation happens, len(curValidator) > len(snapshot)
+	// if it doesn't happen, curValidator == snapshot
+	// if there are only removals, curValidator < snapshot
+
 	// Payout each delegator's reward pro-rata
 	totalRewardForDelegators := big.NewInt(0).Set(rewardPool)
-	for i := range snapshot.Delegations {
-		delegation := snapshot.Delegations[i]
-		percentage, ok := shareLookup[delegation.DelegatorAddress]
+	if !nilDelegationsRemoved {
+		for i := range snapshot.Delegations {
+			delegation := snapshot.Delegations[i]
+			percentage, ok := shareLookup[delegation.DelegatorAddress]
 
-		if !ok {
-			return errors.Wrapf(err, "missing delegation shares for reward distribution")
-		}
+			if !ok {
+				return errors.Wrapf(err, "missing delegation shares for reward distribution")
+			}
 
-		rewardInt := percentage.MulInt(totalRewardForDelegators).RoundInt()
-		curDelegation := curValidator.Delegations[i]
-		curDelegation.Reward.Add(curDelegation.Reward, rewardInt)
-		rewardPool.Sub(rewardPool, rewardInt)
+			rewardInt := percentage.MulInt(totalRewardForDelegators).RoundInt()
+			curDelegation := curValidator.Delegations[i]
+			// should we check that curDelegation.DelegatorAddress == delegation.DelegatorAddress ?
+			// wasn't there originally so I leave it for now
+			curDelegation.Reward.Add(curDelegation.Reward, rewardInt)
+			rewardPool.Sub(rewardPool, rewardInt)
+		}
+	} else {
+		// iterate simply over snapshot delegations
+		// those added later to curValidator are new delegations which do not receive rewards immediately
+		// those removed from curValidator are stale delegations which do not receive rewards anyway
+		for i := 0; i < len(snapshot.Delegations); i++ {
+			delegationFromSnapshot := snapshot.Delegations[i]
+			percentage, ok := shareLookup[delegationFromSnapshot.DelegatorAddress]
+			if !ok {
+				return errors.Wrapf(err, "missing delegation shares for reward distribution")
+			}
+			if percentage.IsZero() { // stale delegation
+				continue
+			}
+			// try to find in wrapper
+			// this needs to be O(N^2) because the order is not guaranteed
+			// for example, snapshot is A / B / C / D / E where C is a stale delegation
+			// wrapper then becomes A / B / D / E
+			// if C then delegates to this validator, the wrapper becomes A / B / D / E / C
+			found := false
+			for j := 0; j < len(curValidator.Delegations) && !found; j++ {
```

Comment:
This could be too costly if the number of delegation in the validator is high (there are some with thousands of delegations). There could be a way to optimize the complexity. Since the two array are only differing by removing some elements, the start index of the searching doesn't have to always be 0, right?

Reply:
My original code (that I did not push out) had that assumption. However, it would not necessarily work in the case where a stale delegation is removed and then added by the delegator (to the same validator) later. Let me think of another idea though because I agree with your opinion.